runtime/common/src/transfer_filter.rs

```diff
@@ -13,32 +13,47 @@
 use cfg_primitives::{AccountId, Balance};
 use cfg_traits::{PreConditions, TransferAllowance};
 use cfg_types::{domain_address::DomainAddress, locations::Location, tokens::CurrencyId};
+use frame_support::{dispatch::TypeInfo, traits::IsSubType};
 use pallet_restricted_tokens::TransferDetails;
 use pallet_restricted_xtokens::TransferEffects;
-use parity_scale_codec::Encode;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
 use sp_core::Hasher;
 use sp_runtime::{
-	traits::{BlakeTwo256, Convert},
+	traits::{BlakeTwo256, Convert, DispatchInfoOf, SignedExtension, StaticLookup},
+	transaction_validity::{InvalidTransaction, TransactionValidityError},
 	DispatchError, DispatchResult, TokenError,
 };
 use xcm::v3::{MultiAsset, MultiLocation};
 
+#[derive(
+	Clone, Copy, PartialOrd, Ord, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen,
+)]
+pub enum FilterCurrency {
+	All,
+	Specific(CurrencyId),
+}
+
 pub struct PreXcmTransfer<T, C>(sp_std::marker::PhantomData<(T, C)>);
 
 impl<
-		T: TransferAllowance<AccountId, CurrencyId = CurrencyId, Location = Location>,
+		T: TransferAllowance<AccountId, CurrencyId = FilterCurrency, Location = Location>,
 		C: Convert<MultiAsset, Option<CurrencyId>>,
 	> PreConditions<TransferEffects<AccountId, CurrencyId, Balance>> for PreXcmTransfer<T, C>
 {
 	type Result = DispatchResult;
 
 	fn check(t: TransferEffects<AccountId, CurrencyId, Balance>) -> Self::Result {
-		let currency_based_check = |sender, destination: MultiLocation, currency| {
+		let currency_based_check = |sender: AccountId, destination: MultiLocation, currency| {
 			T::allowance(
-				sender,
+				sender.clone(),
 				Location::XCM(BlakeTwo256::hash(&destination.encode())),
-				currency,
+				FilterCurrency::Specific(currency),
 			)
+			.or(T::allowance(
+				sender,
+				Location::XCM(BlakeTwo256::hash(&destination.encode())),
+				FilterCurrency::All,
+			))
```

Comment:
A couple of points:
 1. I'm pretty sure the right side of the `or()` is computed before checking whether the left side is true. I think you want here `or_else()` which takes a closure.
 2. If the first is false, and then we check for `FilterCurrency::All`. Is it not the same as just checking it for `FilterCurrency::All`?

Reply:
Good point. Will refactor

Reply:
Done in https://github.com/centrifuge/centrifuge-chain/pull/1693/commits/13906f416fa726d553fdc41c5ac7ea70124afcba